wave_lang/kernel/wave/templates/gemm.py

```diff
@@ -339,3 +339,372 @@ def k_loop(acc: tkl.Register[M, N, tkl.f32]) -> tkl.Register[M, N, tkl.f32]:
+            + extra_iter
+        )
+        next_extra_iter = tkw.minimum(cta_id + tkw.scalar(1, i32), sk_extra_iters)
+        work_unit_end = (
+            total_full_tiles * iters_per_output_tile
+            + ((cta_id + tkw.scalar(1, i32)) * sk_iters_pcu)
+            + next_extra_iter
+        )
+
+        tkw.set_symbol(WORK_UNIT_END, work_unit_end)
+        sk_condition = WORK_UNIT_START < WORK_UNIT_END
+
+        @tkw.iterate(
+            WORK_UNIT_START, start=work_unit_start, condition=sk_condition, init_args=[]
+        )
+        def sk_loop():
+            cta_k_start = tkw.self_index(WORK_UNIT_START, i32)
+            remainder = cta_k_start % iters_per_output_tile
+            cta_k_end = tkw.minimum(
+                cta_k_start + (iters_per_output_tile - remainder), work_unit_end
+            )
+            output_tile_id = cta_k_start // iters_per_output_tile
+            tkw.set_symbol(OUTPUT_TILE_ID, output_tile_id)
+            tkw.set_symbol(CTA_K_END, cta_k_end)
+
+            m_offset = (output_tile_id // tkw.scalar(N_TILES, i32)) * tkw.scalar(
+                BLOCK_M, i32
+            )
+            n_offset = (output_tile_id % tkw.scalar(N_TILES, i32)) * tkw.scalar(
+                BLOCK_N, i32
+            )
+            tkw.set_symbol(CTA_M_OFFSET, m_offset)
+            tkw.set_symbol(CTA_N_OFFSET, n_offset)
+
+            tkw.set_symbol(START_K_TILE, remainder)
+            tkw.set_symbol(NUM_K_TILES, cta_k_end - cta_k_start)
+
+            c_reg = Register[M, N, f32](0.0)
+
+            @tkw.iterate(K, init_args=[c_reg])
+            def mac_loop(acc: Register[M, N, f32]) -> Register[M, N, f32]:
+                a_reg = tkw.read(a, mapping=a_read_mapping)
+                b_reg = tkw.read(b, mapping=b_read_mapping)
+                acc = tkw.mma(a_reg, b_reg, acc)
+                return acc
+
+            output_tile_iter_start = output_tile_id * iters_per_output_tile
+            is_first_split = cta_k_start == output_tile_iter_start
+
+            @tkw.conditional(~is_first_split)
+            def store_partial():
+                one_flag = Register[LOCK_DIM, f32](1.0)
+                tkw.write(
+                    mac_loop,
+                    partial_buffer,
+                    mapping=partial_buffer_write_mapping,
+                    volatile=True,
+                )
+                tkw.write(one_flag, lock_buffer, mapping=lock_buffer_write_mapping)
+
+            @tkw.conditional(is_first_split)
+            def reduce_partial():
+                loop_start = cta_id + tkw.scalar(1, i32)
+                output_tile_iter_end = output_tile_iter_start + iters_per_output_tile
+
+                tkw.set_symbol(NEW_CTA_K_END, cta_k_end)
+                new_cta_k_end = tkw.scalar(NEW_CTA_K_END, i32)
+
+                tid = tkw.scalar(THREAD_0, i32)
+                tkw.set_symbol(THREAD_ID, tid)
+
+                # TODO: unnecessary read/write, right now it reads/writes to partial buffer
```

Comment:
lets spend some time together going over the specifics of this. we can invite @willghatch to the call as well.